third_party/tonic/logging/dart_error.cc

```diff
@@ -10,18 +10,66 @@
 namespace tonic {
 namespace DartError {
 const char kInvalidArgument[] = "Invalid argument.";
+DartPersistentValue on_error;
+UnhandledExceptionReporter log_unhandled_exception = [](const std::string&,
+                                                        const std::string&) {};
+
+void ReportUnhandledException(Dart_Handle exception_handle,
+                              Dart_Handle stack_trace_handle) {
+  const std::string exception =
+      tonic::StdStringFromDart(Dart_ToString(exception_handle));
+  const std::string stack_trace =
+      tonic::StdStringFromDart(Dart_ToString(stack_trace_handle));
+  DartError::log_unhandled_exception(exception, stack_trace);
+}
 }  // namespace DartError
 
-bool LogIfError(Dart_Handle handle) {
+void SetErrorHandler(DartState* dart_state, Dart_Handle value) {
+  DartError::on_error.Clear();
+  if (Dart_IsNull(value)) {
+    return;
+  }
+  DartError::on_error.Set(dart_state, value);
+}
+
+void SetUnhandledExceptionReporter(
+    DartError::UnhandledExceptionReporter reporter) {
+  DartError::log_unhandled_exception = reporter;
+}
+
+bool CheckAndHandleError(Dart_Handle handle) {
+  // Specifically handle UnhandledExceptionErrors first. These exclude fatal
+  // errors that are shutting down the vm and compilation errors in source code.
   if (Dart_IsUnhandledExceptionError(handle)) {
     Dart_Handle exception_handle = Dart_ErrorGetException(handle);
-    const std::string exception =
-        tonic::StdStringFromDart(Dart_ToString(exception_handle));
     Dart_Handle stack_trace_handle = Dart_ErrorGetStackTrace(handle);
-    const std::string stack_trace =
-        tonic::StdStringFromDart(Dart_ToString(stack_trace_handle));
-    tonic::Log("Dart Unhandled Exception: %s, stack trace: %s",
-               exception.c_str(), stack_trace.c_str());
+    Dart_Handle on_error = DartError::on_error.Get();
+
+    // Hooks.dart will call the error handler on PlatformDispatcher if it is
+    // not null. If it is null, returns false, fall into the !handled branch
+    // below and log.
+    // If it is not null, defer to the return value of that closure
+    // to determine whether to report via logging.
+    bool handled = false;
+    if (on_error) {
```

Comment:
Do you have to do a null check on this?

Reply:
I'm assuming it's not safe to call `Dart_InvokeClosure(nullptr, ...)` - and even if it is, why bother doing the work instead of just forwarding the error on to the next thing that can handle it?

Reply:
(at this point, on_error refers to a Dart_Handle, not a a DartPersistentValue reference)

Reply:
Sorry, I meant `Dart_IsNull`.

Reply:
I'm pretty sure `Dart_IsNull` is also not safe to call with nullptr but I can add it as an extra check.

Reply:
`_onError` is declared nullable that's why I ask, so `on_error` could be null.

Reply:
_onError is actually not nullable, but I don't think having an extra null check in here hurts anything.

Reply:
Am I looking at the wrong thing?

https://github.com/flutter/engine/pull/32078/files#diff-57d6953e215d0e5dd7260ee60b665c812aa730566ef0b30f7ff1e3d661143585R1025

Reply:
Yes. It's actually looking up https://github.com/flutter/engine/pull/32078/files#diff-c105d828b82bd42643a76afa9e91867d72c3d28d6ddd4423392c1f2f4de0ea51R119